File: integrations/claude-code/choux_agent_state.py
#!/usr/bin/env python3
"""Report Claude Code's activity to Choux as a `choux.agent.state` event."""

import http.client
import json
import os
import socket
import subprocess
import sys
import time
from typing import Any
from urllib.parse import urlsplit


TIMEOUT_SECONDS = 0.5
DETAIL_LENGTH = 256

class DeliveryFailed(Exception):
    pass


DETAIL_KEYS = {
    "Bash": "command",
    "Write": "file_path",
    "Edit": "file_path",
    "Read": "file_path",
    "Glob": "pattern",
    "Grep": "pattern",
    "Task": "description",
    "WebFetch": "url",
}


def endpoint_target(value: str) -> tuple[str | None, str, int, str, str]:
    if value.startswith("http+unix:"):
        rest = value[len("http+unix:"):]
        separator = rest.rfind(":/")
        if separator < 0:
            raise ValueError("invalid endpoint")
        return rest[:separator], "localhost", 80, "http", rest[separator + 1:]

    parsed = urlsplit(value)
    path = parsed.path + (f"?{parsed.query}" if parsed.query else "")
    port = parsed.port or (443 if parsed.scheme == "https" else 80)
    return None, parsed.hostname or "localhost", port, parsed.scheme, path


def publish(endpoint: str, data: dict[str, Any]) -> None:
    socket_path, host, port, scheme, path = endpoint_target(endpoint)
    body = json.dumps({"type": "choux.agent.state", "data": data, "request": False})

    if socket_path is None and scheme == "https":
        connection: http.client.HTTPConnection = http.client.HTTPSConnection(host, port, timeout=TIMEOUT_SECONDS)
    else:
        connection = http.client.HTTPConnection(host, port, timeout=TIMEOUT_SECONDS)

    if socket_path is not None:
        unix = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        unix.settimeout(TIMEOUT_SECONDS)
        unix.connect(socket_path)
        connection.sock = unix

    try:
        connection.request("POST", path, body, {"content-type": "application/json"})
        response = connection.getresponse()
        status = response.status
        response.read()
        if not 200 <= status < 300:
            raise DeliveryFailed(f"endpoint answered {status}")
    finally:
        connection.close()


def tmux_endpoint() -> str | None:
    if not os.environ.get("TMUX"):
        return None
    try:
        output = subprocess.run(
            ["tmux", "show-environment", "PTYS_EVENT_ENDPOINT"],
            check=False, capture_output=True, text=True, timeout=2,
        ).stdout.strip()
    except (OSError, subprocess.SubprocessError):
        return None
    return output.split("=", 1)[1] if output.startswith("PTYS_EVENT_ENDPOINT=") else None


def endpoint_candidates() -> list[str]:
    """A tmux pane outlives the server that spawned it, so the inherited endpoint
    can point at a dead URL. Tmux tracks the live value - prefer it inside tmux."""
    ordered = [tmux_endpoint(), os.environ.get("PTYS_EVENT_ENDPOINT")]
    return list(dict.fromkeys(value for value in ordered if value))


def clipped(value: object) -> str | None:
    if not isinstance(value, str) or not value.strip():
        return None
    text = value.strip()
    return text if len(text) <= DETAIL_LENGTH else f"{text[:DETAIL_LENGTH]}…"


def detail_for(tool_name: object, tool_input: object) -> str | None:
    if not isinstance(tool_name, str) or not isinstance(tool_input, dict):
        return None
    key = DETAIL_KEYS.get(tool_name)
    return None if key is None else clipped(tool_input.get(key))


def state_for(request: dict[str, Any]) -> dict[str, Any]:
    data: dict[str, Any] = {
        "agent": "claude-code",
        "event": request.get("hook_event_name"),
        "at": int(time.time() * 1000),
        "pid": os.getpid(),
    }
    for key, value in (
        ("pane", os.environ.get("TMUX_PANE")),
        ("cwd", request.get("cwd")),
        ("agentSessionId", request.get("session_id")),
        ("tool", request.get("tool_name")),
        ("detail", detail_for(request.get("tool_name"), request.get("tool_input"))),
        ("message", request.get("message")),
    ):
        if isinstance(value, str) and value:
            data[key] = value
    return data


def main() -> None:
    if not os.environ.get("PTYS_EVENT_ENDPOINT") and not os.environ.get("TMUX"):
        return
    request = json.load(sys.stdin)
    if not isinstance(request, dict) or not isinstance(request.get("hook_event_name"), str):
        return
    data = state_for(request)

    for endpoint in endpoint_candidates():
        try:
            publish(endpoint, data)
            return
        except (OSError, ValueError, http.client.HTTPException, DeliveryFailed):
            continue


if __name__ == "__main__":
    try:
        main()
    except Exception:
        pass
    sys.exit(0)

File: integrations/opencode/choux-agent-state.js
const POST_TIMEOUT_MS = 500;
const DETAIL_LENGTH = 256;
const TMUX_ENDPOINT_TTL_MS = 30_000;

const TOOL_NAMES = {
  bash: "Bash",
  edit: "Edit",
  write: "Write",
  read: "Read",
  list: "List",
  glob: "Glob",
  grep: "Grep",
  task: "Task",
  patch: "Patch",
  skill: "Skill",
  webfetch: "WebFetch",
  websearch: "WebSearch",
  todowrite: "TodoWrite",
  todoread: "TodoRead",
};

const DETAIL_KEYS = {
  Bash: "command",
  Edit: "filePath",
  Write: "filePath",
  Read: "filePath",
  List: "path",
  Glob: "pattern",
  Grep: "pattern",
  Task: "description",
  Patch: "filePath",
  Skill: "name",
  WebFetch: "url",
  WebSearch: "query",
};

function toolNameFor(tool) {
  return typeof tool === "string" && tool ? TOOL_NAMES[tool] ?? tool : undefined;
}

function clipped(value) {
  if (typeof value !== "string" || !value.trim()) return undefined;
  const text = value.trim();
  return text.length <= DETAIL_LENGTH ? text : `${text.slice(0, DETAIL_LENGTH)}…`;
}

function detailFor(toolName, args) {
  if (toolName === undefined || args === null || typeof args !== "object") return undefined;
  const key = DETAIL_KEYS[toolName];
  return key === undefined ? undefined : clipped(args[key]);
}

async function readTmuxEndpoint() {
  try {
    const child = Bun.spawn(["tmux", "show-environment", "PTYS_EVENT_ENDPOINT"], { stdout: "pipe", stderr: "ignore" });
    if (await child.exited !== 0) return undefined;
    const line = (await new Response(child.stdout).text()).trim();
    return line.startsWith("PTYS_EVENT_ENDPOINT=") ? line.slice("PTYS_EVENT_ENDPOINT=".length) : undefined;
  } catch {
    return undefined;
  }
}

let tmuxTracked = { value: undefined, at: 0 };

async function tmuxEndpoint() {
  if (!process.env.TMUX) return undefined;
  const now = Date.now();
  if (now - tmuxTracked.at < TMUX_ENDPOINT_TTL_MS) return tmuxTracked.value;
  tmuxTracked = { value: await readTmuxEndpoint(), at: now };
  return tmuxTracked.value;
}

/** A tmux pane outlives the server that spawned it, so the inherited endpoint can
 * point at a dead URL. Tmux tracks the live value - prefer it inside tmux. */
async function endpointCandidates() {
  const ordered = [await tmuxEndpoint(), process.env.PTYS_EVENT_ENDPOINT];
  return [...new Set(ordered.filter((value) => typeof value === "string" && value.length > 0))];
}

async function post(endpoint, body) {
  const init = {
    method: "POST",
    headers: { "content-type": "application/json" },
    body,
    signal: AbortSignal.timeout(POST_TIMEOUT_MS),
  };
  if (!endpoint.startsWith("http+unix:")) return (await fetch(endpoint, init)).ok;

  const rest = endpoint.slice("http+unix:".length);
  const separator = rest.lastIndexOf(":/");
  if (separator < 0) return false;
  const response = await fetch(`http://localhost${rest.slice(separator + 1)}`, { ...init, unix: rest.slice(0, separator) });
  return response.ok;
}

async function publish(data) {
  const body = JSON.stringify({ type: "choux.agent.state", data, request: false });
  for (const endpoint of await endpointCandidates()) {
    try {
      if (await post(endpoint, body)) return;
    } catch {
      continue;
    }
  }
}

/** An approved request leaves the tool it guards running, so put that back on the badge -
 * except for `task`, whose subagent the reporter already counted when it started. */
function resumedEvent(asked) {
  return asked?.tool === "Task" ? "PostToolUse" : "PreToolUse";
}

export const ChouxAgentState = async ({ directory }) => {
  const asked = new Map();

  const report = (state) => {
    void publish({
      agent: "opencode",
      at: Date.now(),
      pid: process.pid,
      pane: process.env.TMUX_PANE,
      cwd: directory,
      ...state,
    }).catch(() => undefined);
  };

  const permissionAsked = (request) => {
    const tool = toolNameFor(request.permission ?? request.type);
    const callId = request.tool?.callID ?? request.callID;
    const patterns = Array.isArray(request.patterns) ? request.patterns : [request.pattern];
    const detail = clipped(request.title) ?? detailFor(tool, request.metadata) ?? clipped(patterns[0]);
    if (typeof callId === "string" && callId) {
      asked.set(request.id, { callId, sessionId: request.sessionID, tool, detail });
    }
    report({ event: "PermissionRequest", agentSessionId: request.sessionID, toolUseId: callId, tool, detail });
  };

  /** Answered in OpenCode's own dialog instead of ours - report the tool call settled
   * so Choux withdraws the question it still has open for it. */
  const permissionReplied = (properties) => {
    const requestId = properties.requestID ?? properties.permissionID;
    const request = asked.get(requestId);
    asked.delete(requestId);
    const rejected = (properties.reply ?? properties.response) === "reject";
    report({
      event: rejected ? "PermissionDenied" : resumedEvent(request),
      agentSessionId: properties.sessionID,
      toolUseId: request?.callId,
      ...(rejected ? {} : { tool: request?.tool, detail: request?.detail }),
    });
  };

  /** A run can end with a request never answered - drop what was tracked for it. */
  const sessionEnded = (sessionId, event) => {
    for (const [requestId, request] of asked) {
      if (request.sessionId === sessionId) asked.delete(requestId);
    }
    report({ event, agentSessionId: sessionId });
  };

  return {
    event: async ({ event }) => {
      try {
        if (event.type === "permission.asked" || event.type === "permission.updated") permissionAsked(event.properties);
        else if (event.type === "permission.replied") permissionReplied(event.properties);
        else if (event.type === "session.idle" || event.type === "session.error") sessionEnded(event.properties.sessionID, "Stop");
        else if (event.type === "session.deleted") sessionEnded(event.properties.sessionID ?? event.properties.info?.id, "SessionEnd");
      } catch {
        // Never let reporting disturb the agent.
      }
    },
    "chat.message": async ({ sessionID }) => {
      report({ event: "UserPromptSubmit", agentSessionId: sessionID });
    },
    "tool.execute.before": async ({ tool, sessionID, callID }, output) => {
      const toolName = toolNameFor(tool);
      report({
        event: "PreToolUse",
        agentSessionId: sessionID,
        toolUseId: callID,
        tool: toolName,
        detail: detailFor(toolName, output?.args),
      });
    },
    "tool.execute.after": async ({ tool, sessionID, callID }) => {
      const toolName = toolNameFor(tool);
      report({
        event: toolName === "Task" ? "SubagentStop" : "PostToolUse",
        agentSessionId: sessionID,
        toolUseId: callID,
        tool: toolName,
      });
    },
    "experimental.session.compacting": async ({ sessionID }) => {
      report({ event: "PreCompact", agentSessionId: sessionID });
    },
  };
};

export default ChouxAgentState;
